lib/luaxtable.c

```diff
@@ -32,48 +14,170 @@
 #include <lauxlib.h>
 #include <lunatik.h>
 
+#include "luadata.h"
+#include "luarcu.h"
+#include "luaxtable.h"
+
 typedef enum luaxtable_type_e {
 	LUAXTABLE_TMATCH,
 	LUAXTABLE_TTARGET,
 } luaxtable_type_t;
 
 typedef struct luaxtable_s {
 	lunatik_object_t *runtime;
+	lunatik_object_t *skb;
 	union {
 		struct xt_match match;
 		struct xt_target target;
 	};
 	luaxtable_type_t type;
 } luaxtable_t;
 
-static bool luaxtable_match(const struct sk_buff *skb, struct xt_action_param *par)
+static struct {
+	lunatik_object_t *target;
+	lunatik_object_t *match;
+	bool match_fallback;
+	unsigned int target_fallback;
+} luaxtable_hooks = {NULL, NULL, false, XT_CONTINUE};
+
+static int luaxtable_invoke(lua_State *L, luaxtable_t *xtable, const char *op, int nargs, int nret, int fallback)
 {
-	return true;
+	int base = lua_gettop(L) - nargs;
+
+	if (lunatik_getregistry(L, xtable) != LUA_TTABLE) {
+		pr_err("%s: could not find ops table\n", op);
+		goto err;
+	}
+
+	if (lua_getfield(L, -1, op) != LUA_TFUNCTION) {
+		pr_err("%s isn't defined\n", op);
+		goto err;
+	}
+
+	lua_insert(L, base + 1); /* op */
+	lua_pop(L, 1); /* table */
+
+	if (lua_pcall(L, nargs, nret, 0) != LUA_OK) {
+		pr_err("%s error: %s\n", op, lua_tostring(L, -1));
+		goto err;
+	}
+	if (nret == 1) {
+		int ret = luaL_optinteger(L, -1, 0);
+		ret = ret < 0 ? fallback : ret;
+		return ret;
```

Comment:
```suggestion
		int ret = luaL_optinteger(L, -1, 0);
		return ret < 0 ? fallback : ret;
```

what happens if it's returned a positive invalid value? are we handling it somewhere.. what would happen if we simply
```suggestion
		return luaL_optinteger(L, -1, fallback);
```?

what netfilter will do?

Reply:
The return value from target or match is used at many places in the kernel. For each type of table or module (ipv4, ipv6, arp, netlink). For `target` verdicts, it is either compared to a `NF_*` value or first `AND` with `NF_VERDICT_MASK (0x000000ff)` and then compared using `== NF_*`. The behaviour is neither documented nor uniform. It's better we sanitise the positive value also.

Regarding the second example, we already discussed that `fallback` will be set if ret is either `nil` or not of the correct type, but if the module itself returns an illegal value (negative in case `match` which expects bool), it would pass this check.

Reply:
> It's better we sanitise the positive value also.

agreed

Reply:
sanitising using the following condition : 

```c
	if (nret == 1) {
		int ret = luaL_optinteger(L, -1, 0);
		ret = (ret < 0 || ret > NF_MAX_VERDICT) ? fallback : ret;
		return ret;
	}

```